### ISSUE TITLE
min()/max() with default value cause issues with key function type

### ISSUE THREAD

**Issue by pfaion** [python/mypy#17536]
## Bug Report

Hi everyone, thanks for all the work on mypy, I love this tool! 👏 

I run into an issue with this code:

```python
def foo(values: list[dict]) -> dict | None:
    return max(values, default=None, key=lambda entry: entry.get("value", 0))
```
See playground: https://mypy-play.net/?mypy=latest&python=3.12&gist=3daadb92c263a8447595aa7640a04599

### Expected Behavior:
mypy should not report any errors

### Actual Behavior
mypy reports an error:
```
Item "None" of "dict[Any, Any] | None" has no attribute "get"
```

### Environment
- Mypy version used: `mypy 1.10.1 (compiled: yes)`
- Mypy command-line flags: **none**
- Mypy configuration options from `mypy.ini` (and other config files): **none**
- Python version used: `Python 3.12.2`

### Additional Details

I found this issue from 2019, which seems to be the same issue: https://github.com/python/mypy/issues/6460

Here people reported that it was an issue with typeshed, which was fixed in typeshed with this PR: https://github.com/python/typeshed/pull/2833

I checked current trunk for typeshed and cannot spot any issue with the stubs for `max()`, the type correctly seems to indicate that the key function has nothing to do with the type of the default parameter: https://github.com/python/typeshed/blob/d482d4e83c9d247e79c87d954a9ccb11f85bfc37/stdlib/builtins.pyi#L1515-L1516

So while it seems this issue was once fixed it the past, it looks like it might have reappeared through a different cause.

**Comment by pfaion** [comment]
Hmmm... ok, that is unfortunate.

I think if you'd infer the type from the arguments, you might be able to provide something like: `Argument 1 to "foo" has incompatible type "list[int | str]"; expected "list[int]"`. Wouldn't that be somewhat useful?

Or maybe an exception needs to be made for "constructors" in this context? The example you gave sounded like you're interested in errors coming from inner expressions before errors coming from outer expressions. I'm wondering if that really is generally the case, or if that's an exception for when the inner expression is about constructing a type, where that would be more helpful?

I might try understanding the surrounding code a bit more. But thanks for the heads-up, I understand now how tricky this is. I was a bit confused why the type inference was started from the return type, but I thought there must be some explanation behind that!
